File: token-client-core/src/test/java/no/nav/security/token/support/client/core/oauth2/OnBehalfOfTokenClientTest.java
package no.nav.security.token.support.client.core.oauth2;

import no.nav.security.token.support.client.core.ClientProperties;
import no.nav.security.token.support.client.core.OAuth2ClientException;
import no.nav.security.token.support.client.core.OAuth2GrantType;
import no.nav.security.token.support.client.core.http.SimpleOAuth2HttpClient;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.MockitoAnnotations;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

import static no.nav.security.token.support.client.core.TestUtils.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

class OnBehalfOfTokenClientTest {

    private static final String TOKEN_RESPONSE = "{\n" +
        "    \"token_type\": \"Bearer\",\n" +
        "    \"scope\": \"scope1 scope2\",\n" +
        "    \"expires_at\": 1568141495,\n" +
        "    \"ext_expires_in\": 3599,\n" +
        "    \"expires_in\": 3599,\n" +
        "    \"access_token\": \"<base64URL>\",\n" +
        "    \"refresh_token\": \"<base64URL>\"\n" +
        "}\n";

    private static final String ERROR_RESPONSE = "{\"error\": \"some client error occurred\"}";

    private static final String TOKEN_ENDPOINT = "/oauth2/v2.0/token";
    private OnBehalfOfTokenClient onBehalfOfTokenResponseClient;
    private String tokenEndpointUrl;
    private MockWebServer server;

    @BeforeEach
    void setup() throws IOException {
        MockitoAnnotations.initMocks(this);
        this.server = new MockWebServer();
        this.server.start();
        this.tokenEndpointUrl = this.server.url(TOKEN_ENDPOINT).toString();
        onBehalfOfTokenResponseClient = new OnBehalfOfTokenClient(new SimpleOAuth2HttpClient());
    }

    @AfterEach
    void teardown() throws IOException {
        server.shutdown();
    }

    @Test
    void getTokenResponse() throws InterruptedException {
        this.server.enqueue(jsonResponse(TOKEN_RESPONSE));
        String assertion = jwt("sub1").serialize();
        ClientProperties clientProperties = clientProperties(this.tokenEndpointUrl, OAuth2GrantType.JWT_BEARER);
        OnBehalfOfGrantRequest oAuth2OnBehalfOfGrantRequest = new OnBehalfOfGrantRequest(clientProperties, assertion);
        OAuth2AccessTokenResponse response =
            onBehalfOfTokenResponseClient.getTokenResponse(oAuth2OnBehalfOfGrantRequest);

        RecordedRequest recordedRequest = server.takeRequest();
        assertPostMethodAndJsonHeaders(recordedRequest);
        String formParameters = recordedRequest.getBody().readUtf8();
        assertThat(formParameters).contains("grant_type=" + URLEncoder.encode(OAuth2GrantType.JWT_BEARER.value(),
            StandardCharsets.UTF_8));
        assertThat(formParameters).contains("scope=scope1+scope2");
        assertThat(formParameters).contains("requested_token_use=on_behalf_of");
        assertThat(formParameters).contains("assertion=" + assertion);

        assertThat(response).isNotNull();
        assertThat(response.getAccessToken()).isNotBlank();
        assertThat(response.getExpiresAt()).isPositive();
        assertThat(response.getExpiresIn()).isPositive();
    }

    @Test
    void getTokenResponseWithError() {
        this.server.enqueue(jsonResponse(ERROR_RESPONSE).setResponseCode(400));
        String assertion = jwt("sub1").serialize();
        ClientProperties clientProperties = clientProperties(this.tokenEndpointUrl, OAuth2GrantType.JWT_BEARER);
        OnBehalfOfGrantRequest oAuth2OnBehalfOfGrantRequest = new OnBehalfOfGrantRequest(clientProperties, assertion);
        assertThatExceptionOfType(OAuth2ClientException.class)
            .isThrownBy(() -> onBehalfOfTokenResponseClient.getTokenResponse(oAuth2OnBehalfOfGrantRequest))
            .withMessageContaining(ERROR_RESPONSE);
    }
}

File: token-client-core/src/test/kotlin/no/nav/security/token/support/client/core/oauth2/OnBehalfOfTokenClientTest.kt
package no.nav.security.token.support.client.core.oauth2

import com.nimbusds.oauth2.sdk.GrantType.JWT_BEARER
import java.net.URLEncoder
import java.nio.charset.StandardCharsets.UTF_8
import okhttp3.mockwebserver.MockWebServer
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import no.nav.security.token.support.client.core.OAuth2ClientException
import no.nav.security.token.support.client.core.TestUtils.assertPostMethodAndJsonHeaders
import no.nav.security.token.support.client.core.TestUtils.clientProperties
import no.nav.security.token.support.client.core.TestUtils.jsonResponse
import no.nav.security.token.support.client.core.TestUtils.jwt
import no.nav.security.token.support.client.core.http.SimpleOAuth2HttpClient

internal class OnBehalfOfTokenClientTest {

    private lateinit var onBehalfOfTokenResponseClient : OnBehalfOfTokenClient
    private lateinit var tokenEndpointUrl : String
    private lateinit var server : MockWebServer
    @BeforeEach
    fun setup() {
        server = MockWebServer()
        server.start()
        tokenEndpointUrl = server.url(TOKEN_ENDPOINT).toString()
        onBehalfOfTokenResponseClient = OnBehalfOfTokenClient(SimpleOAuth2HttpClient())
    }

    @AfterEach
    fun teardown() {
        server.shutdown()
    }

    @Test
    fun tokenResponse()  {
            server.enqueue(jsonResponse(TOKEN_RESPONSE))
            val assertion = jwt("sub1").serialize()
            val clientProperties = clientProperties(tokenEndpointUrl, JWT_BEARER)
            val oAuth2OnBehalfOfGrantRequest = OnBehalfOfGrantRequest(clientProperties, assertion)
            val response = onBehalfOfTokenResponseClient.getTokenResponse(oAuth2OnBehalfOfGrantRequest)
            val recordedRequest = server.takeRequest()
            assertPostMethodAndJsonHeaders(recordedRequest)
            val formParameters = recordedRequest.body.readUtf8()
            assertThat(formParameters)
                .contains("grant_type=${URLEncoder.encode(JWT_BEARER.value, UTF_8)}")
                .contains("scope=scope1+scope2")
                .contains("requested_token_use=on_behalf_of")
                .contains("assertion=$assertion")
            assertThat(response).isNotNull()
            assertThat(response?.accessToken).isNotBlank()
            assertThat(response?.expiresAt).isPositive()
            assertThat(response?.expiresIn).isPositive()
        }

    @Test
    fun  tokenResponseWithError() {
            server.enqueue(jsonResponse(ERROR_RESPONSE).setResponseCode(400))
            val assertion = jwt("sub1").serialize()
            val clientProperties = clientProperties(tokenEndpointUrl, JWT_BEARER)
            val oAuth2OnBehalfOfGrantRequest = OnBehalfOfGrantRequest(clientProperties, assertion)
            assertThrows<OAuth2ClientException> {
                val res = onBehalfOfTokenResponseClient.getTokenResponse(oAuth2OnBehalfOfGrantRequest)
                println(res)
            }
    }

    companion object {

        private val TOKEN_RESPONSE = """{
             "token_type": "Bearer",
             "scope": "scope1 scope2",
             "expires_at": 1568141495,
             "ext_expires_in": 3599,
             "expires_in": 3599,
             "access_token": "<base64URL>",
             "refresh_token": "<base64URL>"
        }
        """.trimIndent()
        private const val ERROR_RESPONSE = """{"error": "some client error occurred"}"""
        private const val TOKEN_ENDPOINT = "/oauth2/v2.0/token"
    }
}